core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1357,19 +1382,60 @@ public void startRecovery(DiscoveryNode localNode, DiscoveryNode sourceNode, Rec
             }
         } else if (restoreSource == null) {
             // recover from filesystem store
+
+            IndexMetaData indexMetaData = indexSettings().getIndexMetaData();
+            Index mergeSourceIndex = indexMetaData.getMergeSourceIndex();
+            final boolean recoverFromLocalShards = mergeSourceIndex != null && shardRouting.allocatedPostIndexCreate(indexMetaData) == false && shardRouting.primary();
             final RecoveryState recoveryState = new RecoveryState(shardId(), shardRouting.primary(),
-                RecoveryState.Type.STORE, localNode, localNode);
-            markAsRecovering("from store", recoveryState); // mark the shard as recovering on the cluster state thread
-            threadPool.generic().execute(() -> {
-                try {
-                    if (recoverFromStore(localNode)) {
-                        recoveryListener.onRecoveryDone(recoveryState);
+                recoverFromLocalShards ? RecoveryState.Type.LOCAL_SHARDS : RecoveryState.Type.STORE, localNode, localNode);
+            if (recoverFromLocalShards) {
+                final List<IndexShard> startedShards = new ArrayList<>();
+                IndexMetaData sourceIndexMetaData = indicesService.getIndexMetaData(mergeSourceIndex);
```

Comment:
Copying over from the previous iteration - can we get the metaData from IndexService.getIndexSettings? This will mean IndicesService doesn't need to proxy clusterService.state()